src/xenia/kernel/achievement_dispatcher.cc

```diff
@@ -0,0 +1,253 @@
+#include "xenia/kernel/kernel_state.h"
+#include "xenia/kernel/util/shim_utils.h"
+#include "xenia/ui/imgui_guest_notification.h"
+
+DEFINE_bool(show_achievement_notification, false,
+            "Show achievement notification on screen.", "UI");
+
+DEFINE_string(default_achievement_backend, "",
+              "Defines which achievement backend should be used as an defult. "
+              "Possible options: [].",
+              "Achievements");
+
+DECLARE_int32(user_language);
+
+namespace xe {
+namespace kernel {
+
+GpdAchievementBackend::GpdAchievementBackend() { achievements_.clear(); }
+
+GpdAchievementBackend::~GpdAchievementBackend() {}
+
+void GpdAchievementBackend::EarnAchievement(const uint64_t xuid,
+                                            const uint32_t title_id,
+                                            const uint32_t achievement_id) {
+  const auto user = kernel_state()->user_profile(xuid);
+  if (!user) {
+    return;
+  }
+
+  auto achievement = GetAchievementInfoInternal(xuid, title_id, achievement_id);
+  if (!achievement) {
+    return;
+  }
+
+  XELOGI("Player: {} Unlocked Achievement: {}", user->name(),
+         xe::to_utf8(xe::load_and_swap<std::u16string>(
+             achievement->achievement_name.c_str())));
+
+  const uint64_t unlock_time = Clock::QueryHostSystemTime();
+  achievement->flags =
+      achievement->flags | static_cast<uint32_t>(AchievementFlags::kAchieved);
+
+  achievement->unlock_time.high_part = static_cast<uint32_t>(unlock_time << 31);
+  achievement->unlock_time.low_part = static_cast<uint32_t>(unlock_time);
+
+  SaveAchievementData(xuid, title_id, achievement_id);
+}
+
+AchievementGpdStructure* GpdAchievementBackend::GetAchievementInfoInternal(
+    const uint64_t xuid, const uint32_t title_id,
+    const uint32_t achievement_id) const {
+  const auto user = kernel_state()->user_profile(xuid);
+  if (!user) {
+    return nullptr;
+  }
+
+  if (!achievements_.count(xuid)) {
+    return nullptr;
+  }
+
+  for (auto& entry : achievements_.at(xuid)) {
+    if (entry.achievement_id == achievement_id) {
+      return const_cast<AchievementGpdStructure*>(&entry);
+    }
+  }
+
+  return nullptr;
+}
+
+const AchievementGpdStructure* GpdAchievementBackend::GetAchievementInfo(
+    const uint64_t xuid, const uint32_t title_id,
+    const uint32_t achievement_id) const {
+  return GetAchievementInfoInternal(xuid, title_id, achievement_id);
+}
+
+bool GpdAchievementBackend::IsAchievementUnlocked(
+    const uint64_t xuid, const uint32_t title_id,
+    const uint32_t achievement_id) const {
+  const auto achievement =
+      GetAchievementInfoInternal(xuid, title_id, achievement_id);
+
+  return (achievement->flags &
+          static_cast<uint32_t>(AchievementFlags::kAchieved)) != 0;
+}
+
+const std::vector<AchievementGpdStructure*>
+GpdAchievementBackend::GetTitleAchievements(const uint64_t xuid,
+                                            const uint32_t title_id) const {
+  return {};
```

Comment:
IF it is currently running title we can just return what we're storing in a map, however for any other title it should check if there is corresponding file to it and load data from it. When GPD is available